KeysPerSecond/src/me/roan/kps/Main.java

```diff
@@ -272,31 +279,32 @@ protected static final void mainLoop(){
 		future = timer.scheduleAtFixedRate(new Runnable() {
 			@Override
 			public void run() {
-				if(!suspended){
-					int totaltmp = tmp;
-					for(int i : timepoints){
-						totaltmp += i;
-					}
-					if(totaltmp > max){
-						max = totaltmp;
-					}
-					if(totaltmp != 0){
-						avg = (avg * (double)n + (double)totaltmp) / ((double)n + 1.0D);
-						n++;
-						TotPanel.hits += tmp;
-						System.out.println("Current keys per second: " + totaltmp + " time frame: " + tmp);
-					}
-					graph.addPoint(totaltmp);
-					graph.repaint();
-					content.repaint();
-					prev = totaltmp;
-					timepoints.addFirst(tmp);
-					if(timepoints.size() >= 1000 / config.updateRate){
-						timepoints.removeLast();
-					}
-					tmp = 0;
-				}else{
-					tmp = 0;
+				int currtmp = tmp.getAndSet(0);
+
+				if(suspended) return;
+
+				int totaltmp = timepoints.stream().reduce(currtmp, Integer::sum);
+
+				prev = totaltmp;
+				max = Math.max(max, totaltmp);
+
+				if(totaltmp > 0) {
+					total += totaltmp;
+					n++;
+					System.out.println("Current keys per second: " + totaltmp + " time frame: " + currtmp);
+				}
+
+				graph.addPoint(totaltmp);
+				graph.repaint();
+				content.repaint();
+
+				int maxtimepoints = 1000 / config.updateRate - 1;

```

Comment:
This was done so that a new time point object would only be created when it was actually needed, versus being created and immediately removed for the case where updateRate >= 1000.

Reply:
I see. But LinkedList insertion has O(1) as performance right. And since timepoints are Integers and tmp is already instantiated no new object will be created. What would have the best performance executing O(1) twice on an 'empty/size 1' linked list. Or instantiating a new variable, and a loop?
Why did you use a loop actually? At most only one timepoint will have to be removed. If that isn't the case there's a mistake somewhere else.

Reply:
I used the while loop to fix the issue with the number of time points in the array after switching update rates but you've fixed it already in a different way.

Also linked list insertion/removal is in fact O(1); however each int will create a new Integer object in the list because of Autoboxing.

Reply:
Ah ok. I seems that my understand of autoboxing was slightly off.